src/toolkits/drawing_classifier/drawing_classifier.hpp

```diff
@@ -0,0 +1,250 @@
+      "    - 'recall'           : Recall score (macro average)\n"
+      "    - 'f1_score'         : F1 score (macro average)\n"
+      "    - 'log_loss'         : Log loss\n"
+      "    - 'confusion_matrix' : An SFrame with counts of possible\n"
+      "                           prediction/true label combinations.\n"
+      "    - 'roc_curve'        : An SFrame containing information needed for an\n"
+      "                           ROC curve\n"
+  );
+  REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::export_to_coreml,
+                                 "filename");
+
+  REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::load_pretrained_model,
+                                 "model_data", "version");
+
+  END_CLASS_MEMBER_REGISTRATION
+
+ protected:
+  // Override points allowing subclasses to inject dependencies
+
+  // Factory for data_iterator
+  virtual std::unique_ptr<data_iterator> create_iterator(
+      gl_sframe data, bool requires_labels, bool is_train,
+      bool use_data_augmentation) const;
+
+  // Factory for compute_context
+  virtual std::unique_ptr<neural_net::compute_context> create_compute_context()
+      const;
+
+  // Returns the initial neural network to train
+  virtual std::unique_ptr<neural_net::model_spec> init_model() const;
+
+  virtual std::tuple<gl_sframe, gl_sframe> init_data(
+      gl_sframe data, variant_type validation_data) const;
+
+  // Support for iterative training.
+  // TODO: Expose via forthcoming C-API checkpointing mechanism?
+  virtual void init_train(gl_sframe data, std::string target_column_name,
+                          variant_type validation_data,
+                          std::map<std::string, flexible_type> opts);
+  virtual void perform_training_iteration();
+
+  virtual std::tuple<float, float> compute_validation_metrics(
+      size_t num_classes, size_t batch_size);
+
+  virtual void init_table_printer(bool has_validation);
+
+  // Returns an SArray where each entry corresponds to one prediction.
+  virtual gl_sarray perform_inference(data_iterator* data) const;
+
+  // Utility code
+
+  template <typename T>
+  T read_state(const std::string& key) const {
+    return variant_get_value<T>(get_state().at(key));
+  }
+
+ private:
+  // Primary representation for the trained model.
+  std::unique_ptr<neural_net::model_spec> nn_spec_;
+
+  // Primary dependencies for training. These should be nonnull while training
+  // is in progress.
+  gl_sframe training_data_;  // TODO: Avoid storing gl_sframe AND data_iterator.
+  gl_sframe validation_data_;
+  std::unique_ptr<data_iterator> training_data_iterator_;
+  std::unique_ptr<data_iterator> validation_data_iterator_;
+  std::unique_ptr<neural_net::compute_context> training_compute_context_;
+  std::unique_ptr<neural_net::model_backend> training_model_;
+
+  // Nonnull while training is in progress, if progress printing is enabled.
+  std::unique_ptr<table_printer> training_table_printer_;
+};
+
```

Comment:
We don't necessarily have to define anything private or protected in this PR. Let's focus on the interface, and having it callable. In particular, after your PR, we should be able to do the following:
```
import turicreate as tc
model = tc.extensions.drawing_classifier()
model.train(data, 'target', 'drawing') # Shouldn't crash. Might do nothing now, but shouldn't crash.
preds = model.predict(data) # Should return an SArray
```

Reply:
@abhishekpratapa I think this would be a good first step for Style Transfer, too. First get to the point where we can instantiate a (mostly unimplemented) C++ object from Python. Then we can begin developing and testing from Python using the low-level extensions wrapper around C++